AY2021S1-CS2103-T16-3/tp
# Functionality increment: improve help

### What this does
This PR improves how users can get help while using the application, as we have discussed. Closes #42, closes #43.

How help now works:
- Entering `help` shows the user a summary of all the existing commands and a link to the user guide in the results panel on the left.
- Entering `help <command>`  eg. `help rooms`  shows the user a help message for that command in the results panel.
- Pressing the Help button in the menubar or pressing F1 still shows a dialog box, but it now prompts the user to either enter the command `help` or visit the user guide online (the URL has been updated to our team's user guide).

Tests have also been added for:
- AddressBookParser
- HelpCommand

### How to test
- Test the following commands: `help`, `help <any command>` and `help <invalid command>`
- To check the help dialog: click the help button in the menubar or press the F1 key

### Notes
To implement the `help` command, I also made some changes to how commands are added:
- The `MESSAGE_USAGE` constant in each Command has been replaced by a Help object named `HELP` to ensure consistent formatting of help messages.
- To bind a command to the UI, instead of adding the command as a switch case to `AddressBookParser.java`, the (newly created) CommandMapper class is modified as shown below. This class ensures that when Commands are added or removed in the future, the help messages from the help command get updated too, without us having to modify `HelpCommand.java`.

```java
    public CommandMapper() {
        commandMap = new CommandMap();
        commandMap.addCommand(AddCommand.COMMAND_WORD, AddCommand.HELP, new AddCommandParser()::parse);
        commandMap.addCommand(ClearCommand.COMMAND_WORD, ClearCommand.HELP, unused -> new ClearCommand());
        // ....
    }
````

~To be done:~
- ~Tests for HelpCommand~ done
